Fix error caused by beta python versions

cordery opened the pull request:
Resolved an issue where a beta python version in the python_version marker would cause an error.

* Fixes #317

cordery commented:
Also there was an implicit bug where version identifiers containing asterisks would have the asterisks removed even if they were in the middle of the identifier, ex 3.*.5 would become '3.5', *.10 would be come '10'.  I altered the parser to stop when an asterisk was encountered to 3.*.5 becomes 3 and *.10 becomes ''.  Empty version identifiers are caught by the Marker class which will raise an appropriate InvalidSpecifier error. 

If this was not a bug and there was a historical reason for parsing these identifiers that way, then I can remove of course.

matteius commented:
Awesome work @cordery -- thanks for putting this together.  Just trying to think more about the asterisk implications; what I wonder is if the asterisk was used to have partial match like say you want python 3.7.* but I guess in that case you would just go with 3.7 and pyenv should grab the latest release of 3.7.*.   How did you discover or run into asterisk being in the version number?  I think this change makes sense and prevents a subtle edge case because `3.*.5` should never be `3.5` -- and it would be quite silly to pin to a patch version and keep the minor version free to move.  So yeah let's see how this test run plays out.

cordery commented:
Hi matteius, I noticed the asterisk issue when altering the `_tuplize_version` function to fix the unnecessary forced integer conversion of the patch identifier portion of the version string.   

The function originally returned the versions like so: `return tuple(int(x) for x in filter(lambda i: i != "*", version.split(".")))`, the `filter` would drop the asterisks, which would have the effect of altering the position of any identifiers coming after the asterisk, which would 'upgrade' them from patch to minor, minor to major, etc.   

This was probably intended to cleanup the parsing of an unnecessary _trailing_ asterisks in version numbers, ie `3.*` is semantically identical to `3`, as `3.9.*` is to `3.9`.  It may or may not have bitten anyone, as arguably `*.X` and `X.*.X` are invalid identifiers and if appearing were most probably typos, but I felt that given the critical nature of the package it was better to have it behave as correctly as possible even when confronted with messy input.  I added tests to cover those types of messy version strings.
